### PR TITLE
[FLINK-8339] [flip6] Let CustomCommandLine return ClusterDescriptor

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

Instead of directly retrieving or deploying a Flink cluster, the
CustomCommandLine now only returns a ClusterDescriptor which can be used
for these operations. This disentangles the ClusterDescriptor and the
CustomCommandLine a bit better supporting a proper lifecycle management
of the former.

This PR is based on #5224. 

## Brief change log

- Remove indirection that `CustomCommandLines` can retrieve and deploy Flink clusters
- Let `CustomCommandLine` return `ClusterDescriptor`
- Adapt command methods in `CliFrontend` to use the `ClusterDescriptor` instead

## Verifying this change

- Covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java [hunk 1]
```diff
@@ -51,45 +45,23 @@ public String getId() {
...
-			Configuration config,
-			String configurationDirectory,
-			List<URL> userJarFiles) throws UnsupportedOperationException {
-
-		StandaloneClusterDescriptor descriptor = new StandaloneClusterDescriptor(config);
-		ClusterSpecification clusterSpecification = ClusterSpecification.fromConfiguration(config);
-
-		return descriptor.deploySessionCluster(clusterSpecification);
+	public ClusterSpecification getClusterSpecification(Configuration configuration, CommandLine commandLine) {
+		return new ClusterSpecification.ClusterSpecificationBuilder().createClusterSpecification();
```
Comment (GJL) [inline flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java]: Can this code actually be executed? It shouldn't be possible to deploy a standalone cluster from what I understand:
`StandaloneClusterDescriptor#deploySessionCluster(ClusterSpecification clusterSpecification)`
```
	@Override
	public StandaloneClusterClient deploySessionCluster(ClusterSpecification clusterSpecification) throws UnsupportedOperationException {
		throw new UnsupportedOperationException("Can't deploy a standalone cluster.");
	}
```

If this shouldn't run, maybe throw an Exception directly.

File: flink-clients/src/main/java/org/apache/flink/client/cli/Flip6DefaultCLI.java [hunk 2]
```diff
@@ -56,43 +51,30 @@ public String getId() {
...
-		ClusterSpecification clusterSpecification = ClusterSpecification.fromConfiguration(config);
+	@Nullable
+	public String getClusterId(Configuration configuration, CommandLine commandLine) {
+		return "flip6Standalone";
+	}
 
-		return descriptor.deploySessionCluster(clusterSpecification);
+	@Override
+	public ClusterSpecification getClusterSpecification(Configuration configuration, CommandLine commandLine) {
+		return new ClusterSpecification.ClusterSpecificationBuilder().createClusterSpecification();
```
Comment (GJL) [inline flink-clients/src/main/java/org/apache/flink/client/cli/Flip6DefaultCLI.java]: Same thing here. See comment on `DefaultCLI.java`.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 3]
```diff
@@ -457,7 +499,19 @@ protected int stop(String[] args) throws Exception {
 
 		final CustomCommandLine<?> activeCommandLine = getActiveCustomCommandLine(commandLine);
 
-		final ClusterClient client = activeCommandLine.retrieveCluster(commandLine, configuration, configurationDirectory);
+		final ClusterDescriptor<?> clusterDescriptor = activeCommandLine.createClusterDescriptor(
```
Comment (GJL) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: It seems that there is a pattern of getting the `clusterDescriptor` and and the `client` for all the commands. Also shutting down the client and closing the descriptor is repeated. Imo we should extract the pattern to avoid code duplication.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 4]
```diff
@@ -561,7 +640,7 @@ protected int cancel(String[] args) throws Exception {
 	 *
 	 * @param args Command line arguments for the cancel action.
 	 */
-	protected int savepoint(String[] args) throws CliArgsException {
+	protected void savepoint(String[] args) throws Exception {
```
Comment (GJL) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: Why was this changed? It's the only command which doesn't return a status code. The diff would have been smaller without it.

File: flink-clients/src/test/java/org/apache/flink/client/cli/DefaultCLITest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		CommandLine commandLine = defaultCLI.parseCommandLineOptions(args, false);
+
+		final String localhost = "localhost";
+		final int port = 1234;
+		final Configuration configuration = new Configuration();
+
+		configuration.setString(JobManagerOptions.ADDRESS, localhost);
+		configuration.setInteger(JobManagerOptions.PORT, port);
+
+		final InetSocketAddress expectedAddress = new InetSocketAddress(manualHostname, manualPort);
```
Comment (GJL) [inline flink-clients/src/test/java/org/apache/flink/client/cli/DefaultCLITest.java]: nit: variable is declared too early. Will be used at the `assertEquals`.

File: flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyClusterDescriptor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+	private final C clusterClient;
+
+	public DummyClusterDescriptor(C clusterClient) {
+		this.clusterClient = Preconditions.checkNotNull(clusterClient);
+	}
+
+	@Override
+	public String getClusterDescription() {
+		return null;
```
Comment (GJL) [inline flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyClusterDescriptor.java]: Method is not annotated with `@Nullable` in the `ClusterDescriptor` interface.

File: flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyClusterDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		return clusterClient;
+	}
+
+	@Override
+	public C deployJobCluster(ClusterSpecification clusterSpecification, JobGraph jobGraph) {
+		return clusterClient;
+	}
+
+	@Override
+	public void close() throws Exception {
```
Comment (GJL) [inline flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyClusterDescriptor.java]: nit: `throws Exception` not needed. Doesn't really matter either at the moment.

File: flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyCustomCommandLine.java [hunk 8]
```diff
@@ -62,12 +63,21 @@ public void addGeneralOptions(Options baseOptions) {
...
-	public T createCluster(String applicationName, CommandLine commandLine, Configuration config, String configurationDirectory, List<URL> userJarFiles) throws Exception {
-		return clusterClient;
+	@Nullable
+	public String getClusterId(Configuration configuration, CommandLine commandLine) {
+		return "dummy";
+	}
+
+	@Override
+	public ClusterSpecification getClusterSpecification(Configuration configuration, CommandLine commandLine) {
+		return new ClusterSpecification.ClusterSpecificationBuilder().createClusterSpecification();
```
Comment (GJL) [inline flink-clients/src/test/java/org/apache/flink/client/cli/util/DummyCustomCommandLine.java]: See comment on `DefaultCLI.java`

File: flink-clients/src/main/java/org/apache/flink/client/cli/AbstractCustomCommandLine.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * @param <C> type of the ClusterClient which is returned
+ */
+public abstract class AbstractCustomCommandLine<C extends ClusterClient> implements CustomCommandLine<C> {
+
+	protected static final Option ZOOKEEPER_NAMESPACE_OPTION = new Option("z", "zookeeperNamespace", true,
+		"Namespace to create the Zookeeper sub-paths for high availability mode");
+
+
+	protected static final Option ADDRESS_OPTION = new Option("m", "jobmanager", true,
```
Comment (GJL) [inline flink-clients/src/main/java/org/apache/flink/client/cli/AbstractCustomCommandLine.java]: Why are the options here constants? In `FlinkYarnSessionCli` the `Option`s are instance variables. Also `Option`s are mutable.

File: flink-yarn/src/test/java/org/apache/flink/yarn/AbstractYarnClusterTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+			temporaryFolder.newFolder().getAbsolutePath(),
+			yarnClient);
+
+		clusterDescriptor.retrieve(clusterId);
+
+		fail("We should not be able to retrieve ClusterClient for a finished Yarn application.");
+	}
+
+	@Test
+	public void testClusterClientRetrievalFromInvalidYarnId() {
```
Comment (GJL) [inline flink-yarn/src/test/java/org/apache/flink/yarn/AbstractYarnClusterTest.java]: This test is empty.

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+		final Configuration configuration = new Configuration();
+		configuration.setString(YarnConfigOptions.PROPERTIES_FILE_LOCATION, directoryPath.getAbsolutePath());
+
+		final CommandLine commandLine = flinkYarnSessionCli.parseCommandLineOptions(new String[] {}, true);
+
+		flinkYarnSessionCli.getClusterId(
+			configuration,
+			commandLine);
+
+		fail("We should have failed reading the yarn properties file when retrieving the cluster id.");
```
Comment (GJL) [inline flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: It is wrong to put `fail()` if your test expects an exception to be thrown. For example: 
```
@Test(expected = IllegalConfigurationException.class)
	public void testInvalidYarnPropertiesFile() throws Exception {
		try {
			File directoryPath = writeYarnPropertiesFile(invalidPropertiesFile);

			final FlinkYarnSessionCli flinkYarnSessionCli = new FlinkYarnSessionCli("y", "yarn");

			final Configuration configuration = new Configuration();
			configuration.setString(YarnConfigOptions.PROPERTIES_FILE_LOCATION, directoryPath.getAbsolutePath());

			final CommandLine commandLine = flinkYarnSessionCli.parseCommandLineOptions(new String[]{}, true);

			flinkYarnSessionCli.getClusterId(
				configuration,
				commandLine);
		} catch (final Exception e) {

		}

		fail("We should have failed reading the yarn properties file when retrieving the cluster id.");
	}
```

fails with 
```
java.lang.Exception: Unexpected exception, expected<org.apache.flink.configuration.IllegalConfigurationException> but was<java.lang.AssertionError>
```

The message *We should have failed reading the yarn properties file when retrieving the cluster id.* is lost.

File: flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java]: True. Will throw an `UnsupportedOperationException`.

Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java]: After thinking again about it, I think it's ok how it is. The unsupported operation is `StandaloneClusterDescriptor#deploySessionCluster`. But this has nothing to do with the `DefaultCLI#getClusterSpecification`. In the future it might very well be that we can also start a standalone cluster programmatically.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 12]
```diff
@@ -561,7 +640,7 @@ protected int cancel(String[] args) throws Exception {
 	 *
 	 * @param args Command line arguments for the cancel action.
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: good catch

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: I forgot to remove the return codes from the other methods. Will change it.

File: flink-clients/src/test/java/org/apache/flink/client/cli/DefaultCLITest.java (hunk 5 above)
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/cli/DefaultCLITest.java]: Will change it.

File: flink-clients/src/main/java/org/apache/flink/client/cli/AbstractCustomCommandLine.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/AbstractCustomCommandLine.java]: True. Will make it an instance field.

File: flink-yarn/src/test/java/org/apache/flink/yarn/AbstractYarnClusterTest.java (hunk 10 above)
Comment (tillrohrmann) [reply flink-yarn/src/test/java/org/apache/flink/yarn/AbstractYarnClusterTest.java]: Forgot about it. Will add it.

File: flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java (hunk 11 above)
Comment (tillrohrmann) [reply flink-yarn/src/test/java/org/apache/flink/yarn/FlinkYarnSessionCliTest.java]: True. Will change it.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: You're right. I would like to postpone addressing this issue until we've introduce the typed cluster id. Otherwise I fear that it would inflict quite some merge conflicts.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. I've addressed most of your comments and rebased onto the latest master.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 3 above)
Comment (GJL) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: ok

File: flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java (hunk 1 above)
Comment (GJL) [reply flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java]: ok
